### ISSUE TITLE
104 connection reset by peer

### ISSUE THREAD

**Issue by rbrook** [websocket-client/websocket-client#541]
While using the long-live-connection pattern, I occasionally experience hard disconnects saying "Errorno 104, connection reset by peer".
I have multiple WS clients using Gevent monkey patching.
When this happens in one of the WS clients, it crashes the rest of them, and also a similar WS-client running in a separate process. Any idea would be highly appreciated! Thanks.

```Traceback (most recent call last):
  File "/home/rbrook/.virtualenvs/venv/lib/python3.6/site-packages/websocket/_app.py", line 305, in run_forever
    dispatcher.read(self.sock.sock, read, check)
  File "/home/rbrook/projects/wss/socket_helpers.py", line 339, in read
    callback()
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_app.py", line 271, in read
    op_code, frame = self.sock.recv_data_frame(True)
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_core.py", line 343, in recv_data_frame
    frame = self.recv_frame()
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_core.py", line 377, in recv_frame
    return self.frame_buffer.recv_frame()
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_abnf.py", line 361, in recv_frame
    self.recv_header()
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_abnf.py", line 309, in recv_header
    header = self.recv_strict(2)
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_abnf.py", line 396, in recv_strict
    bytes_ = self.recv(min(16384, shortage))
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_core.py", line 452, in _recv
    return recv(self.sock, bufsize)
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_socket.py", line 102, in recv
    bytes_ = _recv()
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/websocket/_socket.py", line 84, in _recv
    return sock.recv(bufsize)
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/gevent/_ssl3.py", line 491, in recv
    return self.read(buflen)
  File "/home/rbrook/.virtualenvs/wss/lib/python3.6/site-packages/gevent/_ssl3.py", line 328, in read
    return self._sslobj.read(len or 1024)
ConnectionResetError: [Errno 104] Connection reset by peer
```

**Comment by rbrook** [comment]
_Update:_
This has probably nothing to do with Gevent.
I managed to easily reproduce the issue without Gevent and when running a single WS client.
When placing `time.sleep(0.01)` after each `ws.send(...)` , the issue seems to be solved. This is very strange because it tells me the socket might not be ready to receive right after sending. 
Perhaps this has something to do with [this discussion](http://www.itmaybeahack.com/homepage/iblog/architecture/C551260341/E20081031204203/index.html)

**Comment by JayDarji** [comment]
Hey @rbrook, I faced this same issue, I only created one Object of websocket and started  run_forever with ping_interval of 45 seconds and ping_timeout of 30 seconds

ws = websocket.WebSocketApp()
ws.run_forever(ping_interval=45, ping_timeout=30) 

I got error of : send_ping routine terminated: [Errno 104] Connection reset by peer

After digging down in websocket library, I got to know that when we specify ping_interval, it creates Daemon thread for sending ping at every interval.

If we got this connection reset error, Is there any way that we can stop existing thread created for ping and recreate new one to solve this problem?

Any Help or Suggestions will be appreciated. Thanks.

**Comment by engn33r** [comment]
If anyone who encounters this can provide more details as listed in [the contributing guidelines](https://github.com/websocket-client/websocket-client/blob/master/CONTRIBUTING.md), I don't have enough information to recreate this issue right now and therefore can't easily fix the issue. Some example code and an example websocket endpoint where this issue can be demonstrated would be helpful.

**Comment by engn33r** [comment]
My current opinion is that this issue is caused by the server and not websocket-client. I experienced a situation where I sent a specific connection request to different websocket servers and got different server responses. Some servers returned a 104 response as describe in this thread, other servers returned a 400 response, etc. Unfortunately error messages are not standardized across websocket server implementations.

For troubleshooting this situation where a strange response status is received, I would recommend examining the server-side logs if you have access to them to understand the server's state at the time that the unexpected status code was sent. If you don't have access to the server logs, you can look at what messages the client sent prior to this error and check for any anomalies (extremely long messages, malformed messages, messages with unexpected values, etc.). You could use [the enableTrace() feature](https://websocket-client.readthedocs.io/en/latest/examples.html#debug-and-logging-options) to capture this information.

**Comment by Tey** [comment]
If ConnectionResetError is raised while reading from the socket, that means the server has closed the connection brutally ("104" is not an HTTP status code, this is a system error code). This should be expected on long lived connections, for instance if the server daemon is restarting or if it kicks idle clients after some time.

Adding `ConnectionResetError` to [this list of exceptions](https://github.com/websocket-client/websocket-client/blob/10d9159af9ba1ecb76e0ebb6a04edfd990eacbbb/websocket/_app.py#L468) should prevent crashing the event loop in this case and trigger a reconnection when allowed.

**Comment by engn33r** [comment]
The ConnectionResetError issue should be resolved by #1048
